pkg/cloud/baremetal/actuators/machine/actuator.go

```diff
@@ -1097,19 +1097,30 @@ func (a *Actuator) remediateIfNeeded(ctx context.Context, machine *machinev1beta
 	}
 
 	if _, needsRemediation := machine.Annotations[externalRemediationAnnotation]; !needsRemediation {
-		return nil
-	}
-
-	node, err := a.getNodeByMachine(ctx, machine)
+		// It can happen that an unhealthy machine with external remediation annotation got healthy, and the external
+		// remediation annotation was removed by the MHC controller, before remediation finished or even started
+		// (e.g. the metal3 pod is not running, but machine was rebooted manually).
+		// In this case stop remediation, if possible.
+
+		// If bmh isn't requested to power off, and the machine isn't marked as powered off for remediation,
+		// there is no ongoing remediation, and there is nothing to do
+		powerOffRequestExists := hasPowerOffRequestAnnotation(baremetalhost)
+		_, poweredOffForRemediationExists := machine.Annotations[poweredOffForRemediation]
+		if !powerOffRequestExists && !poweredOffForRemediationExists {
+			return nil
+		}
 
-	if err != nil {
-		if !errors.IsNotFound(err) {
-			log.Printf("Failed to get Node from Machine %s: %s", machine.Name, err.Error())
-			return err
+		// If bmh was requested to power off, but the machine isn't marked as powered off for remediation yet:
+		// remediation just started and it's fine to stop it by removing the power off request.
+		// That way the bmh won't be powered off at all, or immediately powered on again.
+		if powerOffRequestExists && !poweredOffForRemediationExists {
+			return a.requestPowerOn(ctx, baremetalhost)
```

Comment:
we should check here if the host is still powered on. if it doesn't - it's already too late to stop the remediation. 
Currently this might stop remediation while the node was deleted, and we stuck without a node.

Reply:
the node can't be deleted yet, because the poweredOffForRemediation is set before that step. The only thing which might have happened is that the machine was switched off. Question is: is it ok to just switch it on again, and that's it?

Reply:
> the node can't be deleted yet, because the poweredOffForRemediation is set before that step

That's not what I understand from the code.
Node is deleted at [line 1135](https://github.com/openshift/cluster-api-provider-baremetal/blob/master/pkg/cloud/baremetal/actuators/machine/actuator.go#L1135) and the poweredOffForRemediation annotation is added AFTER it, in [line 1140](https://github.com/openshift/cluster-api-provider-baremetal/blob/master/pkg/cloud/baremetal/actuators/machine/actuator.go#L1140)

Reply:
meh, I should revisit the complete code after I rebased.... thanks for the heads up

Reply:
updated